### ISSUE TITLE
How to build from source?

### ISSUE THREAD

**Issue by ErichDonGubler** [GravityScore/jni-rs#4]
It's unclear how to build from source at the moment. Attempting to build from a Windows 10 box for the stable MSVC target with no attempt at special configuration yields:

```
~/…/experiments/rjni master$ cargo build --example instance
    Updating registry `https://github.com/rust-lang/crates.io-index`
   Compiling libc v0.2.40
   Compiling rjni v0.0.1 (file:///…/experiments/rjni)
error: linking with `link.exe` failed: exit code: 1181
  |
  = note: // snip, horribly long linker invocation
  = note: LINK : fatal error LNK1181: cannot open input file 'jvm.lib'
```

I understand that this project is a work-in-progress, and this may not be an immediate priority for you. However, adding documentation like this will enable people to experiment and contribute, if it's something you think might help this project along. :)

**Comment by benanders** [comment]
Thanks for taking an interest in the project in its current form. You're right, it's not a huge priority for me right now. 

RJNI uses the Java Native Interface to create a Java virtual machine from within Rust, and then interacts with that virtual machine. The problem with this is that Oracle has removed support for creation of a JavaVM using the Java Native Interface in the latest versions of Java. You'll need a legacy Java version (Java 6 I think?), where the `JNI_Create_JavaVM` function is just deprecated, rather than removed. You can find the API documentation for the latest JNI version [here](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html), if you're curious.

In the latest versions of Java, the Java Native Interface is intended to be used when you've got a Java program, then run some Rust from within that Java code, and then need to interact with Java objects from within that Rust code. It seems like it's not meant to create a Java VM from within Rust. So on the whole, this makes me not very motivated either to keep the project up to date, or provide any additional documentation on how to get it working. 

If you're still interested in using it, then you'll need to download a legacy version of Java. The equivalent for macOS can be downloaded from [here](https://support.apple.com/kb/DL1572?locale=en_AU). I don't run a Windows machine so I'm not sure what you'd have to do to get it working.

**Comment by benanders** [comment]
Doing a bit more research, I came across this stack overflow that might help. The person is having the exact same problem as you, it seems: https://stackoverflow.com/questions/28947494/how-to-create-java-vm-in-c?rq=1

**Comment by benanders** [comment]
Thanks for taking an interest. You're right, this project isn't a huge priority for me. I unfortunately don't have a Windows box to experiment on at the moment.

For some reason, macOS requires a legacy version of Java (Java 6) in order for the JNI to work. Maybe you could try downloading the equivalent for Windows? Other than that, I'm out of ideas really. Sorry.

**Comment by joshbooks** [comment]
> macOS requires a legacy version of Java (Java 6) in order for the JNI to work.

pretty sure that's not right, I recently (last job) built a java application for mac with jni components using java 8 (built the app using javafx though, which does a bunch of weird stuff)

**Comment by benanders** [comment]
Yeah all the API calls that RJNI uses are documented in the Java 8 JNI documentation. So I'm sure there's a way to get the project working with Java 8, but for some reason every time I try and use RJNI macOS asks me to download a legacy Java version. Probably something to do with how I'm building the `.class` files and instantiating the JVM.

**Comment by joshbooks** [comment]
definitely to do with how you're instantiating the jvm, if you double click a `.class` file or some equivalent mac pops up that stupid message

**Comment by joshbooks** [comment]
I think the first way I got it to work outside Intellij was to build an app with platypus since apple doesn't like for any java thing to be the primary target of a `.app`

**Comment by ErichDonGubler** [comment]
Were the calls for instantiating the JVM changed or removed but not JNI from Java itself? That would make sense to me.

**Comment by joshbooks** [comment]
[this](https://stackoverflow.com/questions/30695291/linking-with-a-windows-library-outside-the-build-folder) stack overflow post gets it to build (if you make sure you have the right target for your machine) and adding the location of jvm.dll to PATH gets it to run (usually jre/bin/server)

**Comment by joshbooks** [comment]
so I inserted this into my cargo config (%USERPROFILE%/.cargo/config)
```
[target.x86_64-pc-windows-msvc.jvm]
rustc-link-lib = ["jvm"]
rustc-link-search = ["C:/Program Files/Java/jdk1.8.0_161/lib"]
root = "C:/Program Files/Java/jdk1.8.0_161/lib"
```

changed the Cargo.toml to this

```
[package]
name = "rjni"
version = "0.0.1"
authors = ["Ben Anderson <gravityscore@gmail.com>"]
links = "jvm"
build = "src/build.rs"

[dependencies]
libc = "*"

```
added `C:\Program Files\Java\jre1.8.0_161\bin\server` to my PATH env variable, and now it runs on my windows machine, you'll probably need to change some of those paths
